collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java

```diff
@@ -16,79 +16,74 @@
+
     private final WaitingForDataRegistry waitingForDataRegistry;
+    private final CameraUtils cameraUtils;
 
-    public BarcodeWidget(Context context, QuestionDetails questionDetails, WaitingForDataRegistry waitingForDataRegistry) {
+    public BarcodeWidget(Context context, QuestionDetails questionDetails, WaitingForDataRegistry waitingForDataRegistry, CameraUtils cameraUtils) {
         super(context, questionDetails);
         this.waitingForDataRegistry = waitingForDataRegistry;
+        this.cameraUtils = cameraUtils;
+    }
 
-        getBarcodeButton = createSimpleButton(getContext(), getFormEntryPrompt().isReadOnly(), getContext().getString(R.string.get_barcode), getAnswerFontSize(), this);
+    @Override
+    protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
+        binding = BarcodeWidgetAnswerBinding.inflate(((Activity) context).getLayoutInflater());
 
-        stringAnswer = getCenteredAnswerTextView(getContext(), getAnswerFontSize());
+        if (prompt.isReadOnly()) {
+            binding.getBarcodeButton.setVisibility(GONE);
+        } else {
+            binding.getBarcodeButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
+            binding.getBarcodeButton.setOnClickListener(v -> onButtonClick());
+        }
+        binding.barcodeAnswerText.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
 
-        String s = questionDetails.getPrompt().getAnswerText();
-        if (s != null) {
-            getBarcodeButton.setText(getContext().getString(
-                    R.string.replace_barcode));
-            stringAnswer.setText(s);
+        String stringAnswer = prompt.getAnswerText();
+        if (stringAnswer != null && !stringAnswer.isEmpty()) {
+            binding.getBarcodeButton.setText(getContext().getString(R.string.replace_barcode));
+            binding.barcodeAnswerText.setText(stringAnswer);
         }
-        // finish complex layout
-        LinearLayout answerLayout = new LinearLayout(getContext());
-        answerLayout.setOrientation(LinearLayout.VERTICAL);
-        answerLayout.addView(getBarcodeButton);
-        answerLayout.addView(stringAnswer);
-        addAnswerView(answerLayout, WidgetViewUtils.getStandardMargin(context));
+
+        return binding.getRoot();
     }
 
     @Override
     public void clearAnswer() {
-        stringAnswer.setText(null);
-        getBarcodeButton.setText(getContext().getString(R.string.get_barcode));
-
+        binding.barcodeAnswerText.setText(null);
+        binding.getBarcodeButton.setText(getContext().getString(R.string.get_barcode));
         widgetValueChanged();
     }
 
     @Override
     public IAnswerData getAnswer() {
-        String s = stringAnswer.getText().toString();
-        if (s.equals("")) {
-            return null;
-        } else {
-            return new StringData(s);
-        }
+        String stringAnswer = binding.barcodeAnswerText.getText().toString();
```

Comment:
I think just answer would be good name it is String so no need to duplicate it in its name.